docs/enhancements/install-strategy.md

```diff
@@ -0,0 +1,196 @@
+spec.stage | string | Enum string for stage of installation. (Initializing, RequirementsNotMet, Provisioning, RetryProvisioning, Complete, Failed) More details in conditions. *TODO*: Should we just work with conditions? Loss of state in disaster recovery?
+spec.clusterMetadata.adminKubeconfigSecretRef | LocalObjectReference | Admin kubeconfig secret for current install attempt underway, set as soon as known. Copied back to Hive automatically once Ready.
+spec.clusterMetadata.adminPasswordSecretRef | LocalObjectReference | Admin password secret for current install attempt underway, set as soon as known. Copied back to Hive automatically once Ready.
+spec.clusterMetadata.clusterID | string | Cluster UUID generated during install, set as soon as known. Copied back to Hive automatically once Ready.
+spec.clusterMetadata.infraID | string | Cluster infra ID generated during install, set as soon as known. Copied back to Hive automatically once Ready.
+spec.manifestsConfigMapRef | LocalObjectReference | ConfigMap containing manifests to inject into the cluster during installation. *TODO*: Hive technically doesn't need this to be there, do we require it to ensure consistent support or leave it up to install strategies?
+
+The following ClusterDeployment fields are not carring over to this interface as they are too specific to current strategy. They can of course be used by specific stratgies, but Hive will not need to assume they are there.
+
+  * SSHPrivateKeySecretRef: Some install strategies may not want to support this.
+  * SSHKnownHosts
+  * InstallEnv: Too specific to openshift-install.
+  * InstallConfigSecretRef: One of the more contentious parts of the current API.
+  * ReleaseImage: Legacy Hive feature, use ClusterImageSet instead.
+
+All conditions should initialize to Unknown status as soon as possible, and then be set to True/False once known.
+
+Condition | Description
+--------- | -----------
+Ready | Important overall condition indicating if the cluster is ready or not. Reason and Message provide as much detail as possible for the user. Message could contain attempt x of y if the install strategy support retries.
+RequirementsMet | True once we are ready to launch the install. False if we hit a problem and can't start. Reasons: ClusterImageSetNotFound, etc.
+Failed | True once an install attempt has failed. Remains true if retries are in play even if another attempt is in progress. Move to False once Ready goes to True. Reason/Message should contain details on what happened if it's possible to detect, otherwise Unknown. *TODO*: Unfortunate naming here, this matches current ProvisionFailed condition, which means a failure has been encountered, not necessarily that we've given up. Failing would be a better term, but it's risky to change the ProvisionFailed condition meaning on ClusterDeployment when we transfer. Thoughts?
```

Comment:
I don't think it would quite work as moving this to Reason we'd lose some granularity we're relying on, normal openshift installs scan the installer logs for strings that indicate specific problems, each of which becomes a reason, and that list changes often. It would be hard to monitor for one of those problems if they appear on the overall Ready condition. Having them on a ProvisionFailed/Failing condition means you can monitor for that to go True to know something is up, and you don't need to know specifics about the Reason to take action.

Reply:
Still a little stuck here.

At the core I think we need Ready, RequirementsMet, and Failed. But again Failed goes true as soon as "an attempt" failed, we didn't necessarily give up. So how does a caller monitor for whether or not we gave up entirely with just these three conditions.

Reply:
Provisioned - we successfully provisioned (possibly with multiple attempts)
Failed - an attempt failed but we might still be retrying
Completed - true if we're done trying. combine with Provisioned/Failed to know if success or failure. (Reason will also indicate)

Thoughts?

Reply:
Ok Attempt #1125 

Installed - True once we've successfully installed. Reasons can indicate current progress, if we're in retries or not, etc.
Failed - A single attempt failed, includes details on detected problems. We might still be retrying.
Stopped - We've either hit Installed=True, or a limit on install attempts, or we hit a single error and do not support install retries.

Clients can monitor for Installed=True for success, and Stopped=True + Installed=False for a full failure, or Failed=True for any partial failure.

Reply:
in the implementation i have found these conditions to be useful.
https://github.com/openshift/hive/pull/1358/files#diff-5b7e8b4c0bcc87a454db41521d292a136a5c26cf3efb29139b37aa9b07e40accR72-R76

RequirementsMet (this becomes ClusterInstallRequirementsMet on clusterdeployment)
Failed (this becomes ClusterInstallFailed on clusterdeployment)
Completed (this becomes ClusterInstallCompleted on clusterdeployment)

I think we can have the Completed include a reason that the install are Completed = False with Reason ReachedMaxAttempts .  Having a Stopped condition does not seem to provide additional value and it only creates more confusion as it's value can be part of the failure (no more attempts and therefore failure) or success (completed is true so we have stopped). Both these can easily be include in Failed and Completed conditions already...

Reply:
We sorted out in scrum, decision was:

Completed = successfully installed
Failed = an attempt failed, we might still be retrying
Stopped = we are no longer working, combine with Completed/Failed to know result